mono/profiler/coverage.c

```diff
@@ -0,0 +1,922 @@
+ *   <method> - Contains data about methods. Can contain any number of <statement> elements
+ *              Attributes:
+ *                 - assembly: The name of the parent assembly
+ *                 - class: The name of the parent class
+ *                 - name: The name of the method, with all it's parameters
+ *                 - filename: The name of the source file containing this method
+ *                 - token
+ *   <statement> - Contains data about IL statements. Has no child elements
+ *                 Attributes:
+ *                    - offset: The offset of the statement in the IL code after the previous
+ *                              statement's offset
+ *                    - counter: 1 if the line was covered, 0 if it was not
+ *                    - line: The line number in the parent method's file
+ *                    - column: The column on the line
+ */
+
+#include <config.h>
+#include <glib.h>
+
+#include <stdio.h>
+
+#ifdef HAVE_DLFCN_H
+#include <dlfcn.h>
+#endif
+#include <fcntl.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#ifdef HAVE_SYS_MMAN_H
+#include <sys/mman.h>
+#endif
+#if defined (HAVE_SYS_ZLIB)
+#include <zlib.h>
+#endif
+
+#include <mono/metadata/assembly.h>
+#include <mono/metadata/debug-helpers.h>
+#include <mono/metadata/profiler.h>
+#include <mono/metadata/tabledefs.h>
+#include <mono/metadata/metadata-internals.h>
+
+#include <mono/utils/atomic.h>
+#include <mono/utils/hazard-pointer.h>
+#include <mono/utils/lock-free-queue.h>
+#include <mono/utils/mono-conc-hashtable.h>
+#include <mono/utils/mono-os-mutex.h>
+#include <mono/utils/mono-logger-internals.h>
+#include <mono/utils/mono-counters.h>
+
+// Statistics for profiler events.
+static gint32 coverage_methods_ctr,
+              coverage_statements_ctr,
+              coverage_classes_ctr,
+              coverage_assemblies_ctr;
+
+struct _MonoProfiler {
+	MonoProfilerHandle handle;
+
+	FILE* file;
+
+	char *args;
+
+	mono_mutex_t mutex;
+	GPtrArray *data;
+
+	GPtrArray *filters;
+	MonoConcurrentHashTable *filtered_classes;
+	MonoConcurrentHashTable *suppressed_assemblies;
+
+	MonoConcurrentHashTable *methods;
+	MonoConcurrentHashTable *assemblies;
+	MonoConcurrentHashTable *classes;
+
+	MonoConcurrentHashTable *image_to_methods;
+
+	guint32 previous_offset;
+};
+
+typedef struct {
+	//Where to compress the output file
+	gboolean use_zip;
+
+	//Name of the generated xml file
+	const char *output_filename;
+
+	//Filter files used by the code coverage mode
+	GPtrArray *cov_filter_files;
+} ProfilerConfig;
+
+static ProfilerConfig coverage_config;
+static struct _MonoProfiler coverage_profiler;
+
+/* This is a very basic escape function that escapes < > and &
+   Ideally we'd use g_markup_escape_string but that function isn't
+	 available in Mono's eglib. This was written without looking at the
+	 source of that function in glib. */
+static char *
+escape_string_for_xml (const char *string)
```

Comment:
Could we reuse the mprof-report version

Reply:
the end goal is to remove it from `log.c` and `mprof-report.c` so I don't mind the duplication for now.